README.rst
Axiom
=====
Axiom produces a report that can be used to verify if TRIRIGA-generated PM
Schedules are correct. Our customers have noticed that tririga occasionally
creates *duplicates*, *skips scheduled tasks* and otherwise acts inconsistently
when creating tasks from PM Schedules. Axiom can be used to quickly check for
any anomalies.

Getting Started
---------------
Windows Users
~~~~~~~~~~~~~
An installer is available for Windows. Download it from the `Releases
<https://github.com/nithinphilips/axiom/releases>`_ section.  Once installed,
you will have the ``axiom`` command available in your Command Prompt.

All Other Platforms
~~~~~~~~~~~~~~~~~~~
Install Python 3.4. Download the source and install Axiom and its
dependencies::

    python setup.py install

Usage
-----
Simplest use case::

    axiom

This will run the tool against ``localhost:1521:xe`` with all the default
options. If there are no errors, Axiom will not produce any output on the
console. It will create a file (see `The Excel File`_) called ``axiom.xlsx``
in the current directory.

To specify a database::

    axiom --db-url=tridata/tridata@remotedb:1521:orcl --site-url=http://localhost:9080

You can report on a selected set of PM Schedules by passing in the PM Schedule ID::

    axiom --db-url=tridata/tridata@remotedb:1521:orcl \
          --site-url=http://localhost:9080 \
          1000001 1000002

To see all available arguments, run::

    axiom --help

The Excel File
--------------
The output of Axiom is an Excel file. By default it is named ``axiom.xlsx``.
You may change the file name using the ``--outputfile`` (``-f``) flag.

The excel file will have 2 or more worksheets. The first worksheet is always
called *Index*. This sheet summarizes the individual PM Schedules. You can use
the links in the Index worksheet to see more details about individual PM
Schedules.

The **Valid** column shows the number of tasks that Axiom has identified as
correctly dated. **Invalid** column shows the tasks that are not dated
correctly. If the *Valid* count is equal to *Total* count, the *Valid*,
*Invalid* and *Total* cells will be colored green.

The **Task Grouping?** column will indicate whether this PM Schedule is
configured to create a single task for all assets/locations or multiple tasks.
If Task Grouping is Yes, Axiom will group the tasks by Planned Start Date.  You
can see the number of tasks with the same Planned Start date by looking at the
**Task Count** column in the detail worksheet.

Click on the *View* link to open the PM Schedule in TRIRIGA. This will only
work if you specified the site url with the ``--site-url`` (``-u``) flag.

To see detailed validation results for a PM Schedule, click on the *Details*
link. This will take you to the appropriate detail worksheet. This sheet will
display all the tasks associated to the PM Schedule. The **Planned Start Date**
from the work task is compared against an **Expected Date** to see if the
generated task has the correct start date. The results are displayed in the
**Valid?** column.

The **Expected Date** is generated by independently interpreting the PM Event
record associated to a PM Schedule. We use the `python-dateutil
<https://dateutil.readthedocs.org/en/latest/rrule.html>`_ library to turn the
recurrence rules into a series of dates. All the dates are handled in a time-zone
aware manner. By default the time-zone is *US/Eastern*. Use the ``--timezone``
(``-z``) flag to change it to your locale. For USA, *US/Pacific*, *US/Mountain*
and *US/Central* are the common ones.

If your environment has mixed timezones and the time offsets are causing
issues, you can omit the time component by specifying the ``--strip-time``
(``-s``) flag. This way only the dates are compared.

We also apply *Working Calendar Rules* to the expected date. Working calendar
rules are predefined in Axiom. You can choose one using the
``--working-calendar`` (``-w``) flag. Allowed choices are ``8to5`` and
``24/7``. These calendars try to mimic common TRIRIGA calendar rules. *8to5*
will cause any dates before 8 AM local time to be changed to 8 AM. Any dates
after 5 PM will be changed to 8 AM the next day. Note that the Axiom calendar
does not handle the 1 hour lunch period present in TRIRIGA's ``DEFAULT``
calendar. You can only choose a single working calendar per invocation.

The ``8to5`` working calendar rules do not allow tasks to be scheduled on
weekends. The Planned Start Dates that fall on Saturdays or Sundays will be
changed to the following Monday at 8 AM. If you create a DAILY task, you may
notice duplicates because of this. Run Axiom with ``--debug`` and ``-v 2``
flags to see how the dates are changed by the working calendar rules. This
mimics TRIRIGA's behavior.

Use the ``24/7`` calendar to avoid any working calendar restrictions.

Known Issues
------------
* TRIRIGA may not generate a Work Task for the first recurrence in some cases
  because the scheduling engine may not run until after the first the
  recurrence has passed and it designed not to create Work Tasks in the past.
  Axiom does not do this and so, the generated date series could be off by one
  in some cases.
* Axiom uses python-dateutil to generate dates. It implements the iCalendar
  specification. Read `the specification <https://www.ietf.org/rfc/rfc2445.txt>`_
  and `the implementation notes <https://labix.org/python-dateutil>`_ if you
  have questions about Axiom's behavior.
* Working calendar rules are applied after a date is generated from the
  recurrence rules. This could cause dates to duplicate. For example, a DAILY
  schedule will generate recurrences for Sat and Sun. When the working calendar
  rules are applied, the Sat and Sun dates will be changed to the following
  Mon, resulting in 3 tasks for Monday. This is how TRIRIGA behaves with the
  DEFAULT 8 to 5 calendar.
* Axiom does not support shadowing.

Building Windows Installer
--------------------------
Windows installer can be built on Windows machines. You will need Python 3.4
(Windows version) and the ``pyinstaller`` package (version 2.0).

From the project root run::

    make installer

This will build the binaries and installer.

License
-------
.. code::

    Axiom. Verify TRIRIGA PM Schedules.
    Copyright (C) 2016 Nithin Philips

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.


ChangeLog.rst
Change Log
==========
Axiom Change Log

Version 0.2.1
-------------
* FIX: cx_Oracle.InterfaceError: Unable to acquire Oracle environment handle.
  The oracle library is now embedded in the axiom.exe executable.

Version 0.2.0
-------------
* Support for skipping months in MONTHLY schedules
* Support for Ad-hoc schedules.
* Support for *Also Schedule On* dates.
* Support for *Exclude Dates*.
* Users can now specify a list of PMs to report on. Leave blank to report on all PMs.
* Axiom will check associated service plan and automatically group tasks by
  planned start date when the grouping rule is *Create Task For Each
  Asset/Location*. When tasks are grouped the aggregate task *count* is displayed
  in the details table instead of the task ID.
* Add a new column to the details table that compares only the date portion.
* FIX: Resolve issues with timezone being wrong when date is changed by working
  calendar to fall IN or OUT of a DST period.
* FIX: The first task was overwritten by table header.
* FIX: When expected occurrences ended before tasks, an exception was thrown.
  Now Axiom will use a dummy date (Jan 1, 1970) and show all tasks.
* FIX: Do not display error when a PM schedule has no tasks.

Version 0.1.0
-------------
* First release


axiom/pmschedulevalidator.py
#!/cygdrive/c/Python27/python.exe

"""
Produce a report to validate PM Schedules
"""

import logging
import pytz
import xlsxwriter

from argh import arg
from argh.exceptions import CommandError
from datetime import datetime
from tqdm import tqdm
from xlsxwriter.utility import xl_rowcol_to_cell


from .ora_helper import execute, parse_db_url, set_verbosity
from .pmeventparser import get_events, parse_event, restrict_to_working_calendar, localize_date, rrule_str
from .queries import SQL_GET_PMSCHEDS, SQL_GET_PMSCHEDS_FILTERED, SQL_GET_TASKS, SQL_GET_TASKS_GROUPED

xlFormats = dict()

def safe_xl_ws_name(value):
    """
    Normalizes string, converts to lowercase, removes non-alpha characters,
    and converts spaces to hyphens.
    """
    if len(value) >= 32:
        value = value[0:31]
    return value

@arg('pm_id', nargs='*', help="One or more PM Schedule IDs. Leave blank to show all PM Schedules.", default=None)
@arg('-d', '--db-url', help="Database connection string. USERNAME/PASSWORD@HOST:PORT:SID or USERNAME/PASSWORD@HOST:PORT/SERVICE_NAME", default="tridata/tridata@localhost:1521:xe")
@arg('-z', '--timezone', help="The local timezone", default="US/Eastern")
@arg('-u', '--site-url', help="The URL to TRIRIGA. It will be used to generate links to records.", default="http://localhost:9080")
@arg('-f', '--outputfile', help="The output file name.", default="axiom.xlsx")
@arg('-s', '--strip-time', help="Remove time component from dates before comparing.", default=False)
@arg('-w', '--working-calendar', choices=['8to5','24/7'], help="Choose a working calendar", default="8to5")
@arg('-v', '--verbosity', choices=range(0,3), help="Choose how much output to print to console", default=0)
def schedulevalidator(
        pm_id,
        db_url=None,
        timezone=None,
        site_url=None,
        outputfile=None,
        strip_time=None,
        working_calendar=None,
        verbosity=None
    ):
    """
    Produces a report to verify PM Schedules in TRIRIGA.

    Example:

        axiom --db-url=tridata/tridata@localhost:1521:xe --site-url=http://localhost:9080 --timezone=US/Eastern

    """

    set_verbosity(verbosity)

    if site_url.endswith("/"):
        site_url = site_url[:-1]

    db = parse_db_url(db_url)
    logging.debug("Using database connection: " + str(db))
    connection = db.get_connection()

    if pm_id:
        # Get selected PM Schedules
        pms_sql = SQL_GET_PMSCHEDS_FILTERED.format(', '.join(["'{}'".format(i) for i in pm_id]))
        pms = execute(pms_sql, connection)
        if len(pm_id) != len(pms):
            logging.warning("Expected {} PMs, only got {}.".format(len(pm_id), len(pms)))
    else:
        # Get ALL PM Schedules
        pms = execute(SQL_GET_PMSCHEDS, connection)


    if len(pms) <= 0:
        raise CommandError("No PM Schedules found in {}".format(db))

    workbook = xlsxwriter.Workbook(outputfile)
    index_worksheet = workbook.add_worksheet(name="Index")

    xlFormats['bad'] = workbook.add_format({'bg_color': '#FFC7CE', 'font_color': '#9C0006'})
    xlFormats['good'] = workbook.add_format({'bg_color': '#C6EFCE', 'font_color': '#006100'})
    xlFormats['header'] = workbook.add_format({'bold': True, 'text_wrap': True})
    xlFormats['rrule'] = workbook.add_format({'text_wrap': True, 'valign': 'top'})

    # Validate each one
    for pm in tqdm(pms):
        try:
            group = (pm['TRITASKGROUPINGRULELI'] == 'Create Task For Each Asset/Location')

            validate_pm(pm['TRIIDTX'],
                        pm['TRINAMETX'],
                        workbook,
                        connection,
                        timezone,
                        strip_time,
                        working_calendar,
                        group,
                        verbosity)
        except Exception as e:
            logging.exception("Unable to validate a PM. Ignore and continue.")

        if verbosity > 0:
            print()

    populate_index_worksheet(index_worksheet, pms, site_url)
    workbook.close()

def populate_index_worksheet(worksheet, pms, site_url):

    # Widen the columns to make the text clearer.
    worksheet.set_column('A:A', 9)
    worksheet.set_column('B:B', 50)
    worksheet.set_column('C:C', 12)
    worksheet.set_column('D:D', 15)
    worksheet.set_column('E:E', 13)
    worksheet.set_column('F:F', 13)
    worksheet.set_column('G:G', 13)

    for i, pm in enumerate(pms, start=1):
        url = site_url + "/pc/notify/link?recordId=" + str(pm["SPEC_ID"])

        table_name = "Table" + pm["TRIIDTX"]

        worksheet.write(i, 0, pm["TRIIDTX"])
        worksheet.write(i, 1, pm["TRINAMETX"])
        worksheet.write(i, 2, pm["TRIPMTYPECLASSCL"])
        worksheet.write(i, 3, "Yes" if pm["TRITASKGROUPINGRULELI"] == "Create Task For Each Asset/Location" else "No")
        worksheet.write(i, 4, '=COUNTIF({}[Valid?], "OK")'.format(table_name))
        worksheet.write(i, 5, '=COUNTIF({}[Valid?], "ERROR")'.format(table_name))
        formula = "={}+{}".format(xl_rowcol_to_cell(i, 4), xl_rowcol_to_cell(i, 5))
        worksheet.write(i, 6, formula)
        worksheet.write_url(i, 7,
                            url="internal:{}!A1".format(pm["TRIIDTX"]),
                            string="Details",
                            tip="View details about " + pm["TRIIDTX"])
        worksheet.write_url(i, 8,
                            url=url,
                            string="View",
                            tip="View record " + pm["TRIIDTX"] + " in TRIRIGA")

        cellA = xl_rowcol_to_cell(i, 4, row_abs=True, col_abs=True)
        cellB = xl_rowcol_to_cell(i, 6, row_abs=True, col_abs=True)
        worksheet.conditional_format(i, 4, i, 6, {'type': 'formula',
                                                  'criteria': "={}={}".format(cellA, cellB),
                                                  'format': xlFormats['good']})

    worksheet.add_table(0, 0, i, 8, { 'style': 'Table Style Light 11',
                                      'columns': [
                                          {'header': 'PM ID'},
                                          {'header': 'PM Schedule'},
                                          {'header': 'Recurrence'},
                                          {'header': 'Task Grouping?'},
                                          {'header': 'Valid'},
                                          {'header': 'Invalid'},
                                          {'header': 'Total'},
                                          {'header': 'Details'},
                                          {'header': 'View'},
                                     ]})

    worksheet.activate()

def validate_pm(pm_id, pm_name, workbook, connection, timezone, strip_time=False, working_calendar="8to5", grouped=False, verbosity=1):

    event = get_events(connection, pm_id=pm_id, timezone=timezone)

    if len(event) <= 0:
        raise Exception("No recurrence pattern found for PM: {} {}.".format(pm_id, pm_name))
    elif len(event) > 1:
        logging.warning("More than one recurrence patterns found for PM: {} {}. Using the first one.".format(pm_id, pm_name))

    event = event[0]

    rrule, description = parse_event(event, verbosity=verbosity)

    if verbosity >= 1:
        print(pm_id + ": " + pm_name)

    if verbosity >= 2:
        print(description)
        print(str(rrule))

    tasks = execute(SQL_GET_TASKS_GROUPED if grouped else SQL_GET_TASKS,
                    connection,
                    {'pm_id': pm_id })

    if len(tasks) == 1 and tasks[0]['TRIPLANNEDSTARTDT'] is None:
        if verbosity >= 1:
            logging.warning("PM schedule '{}' does not have any tasks.".format(pm_id))
        tasks = [] # Allow a dummy sheet to be created

    total = 0
    ok_count = 0

    local_tz = pytz.timezone(timezone)

    worksheet = workbook.add_worksheet(name=safe_xl_ws_name(pm_id))

    # Widen columns
    worksheet.set_column('A:A', 12)
    worksheet.set_column('B:B', 10)
    worksheet.set_column('C:C', 24)
    worksheet.set_column('D:D', 24)
    worksheet.set_column('F:F', 12)

    worksheet.write(0, 0, pm_id, xlFormats['header'])
    worksheet.merge_range(0, 1, 0, 3, pm_name, xlFormats['header'])
    worksheet.merge_range(1, 0, 1, 3, description, xlFormats['rrule'])
    worksheet.merge_range(2, 0, 2, 3, rrule_str(rrule), xlFormats['rrule'])

    worksheet.write_url(0, 6,
                        url="internal:Index!A1",
                        string="Index",
                        tip="Return to Index page")

    worksheet.set_row(2, 30)

    i = 5 # Sometimes the iteration may not run

    rrule_itr = iter(rrule)
    for i, item in enumerate(tasks, start=5):
        expected_date = next(rrule_itr, datetime.fromtimestamp(0))

        actual_date = localize_date(item['TRIPLANNEDSTARTDT'], local_tz)

        coerced_date = restrict_to_working_calendar(expected_date, working_calendar)
        # The order of execution is important.
        # ``restrict_to_working_calendar`` method may change the date to fall
        # IN or OUT of DST period. This will change the TZ offset.
        expected_date = local_tz.localize(expected_date)
        coerced_date = local_tz.localize(coerced_date)

        if expected_date != coerced_date:
            logging.debug("Coerce date: {} => {}".format(expected_date, coerced_date))

        expected_date = coerced_date

        if strip_time:
            # Strip off time
            expected_date = expected_date.date()
            if actual_date:
                actual_date = actual_date.date()

        ok = expected_date == actual_date

        if ok:
            ok_count = ok_count + 1
        total = total + 1

        formula = '=IF({}={},"OK","ERROR")'.format(xl_rowcol_to_cell(i, 2),
                                                   xl_rowcol_to_cell(i, 3))

        formula2 = '=IF(LEFT({},11)=LEFT({},11),"OK","ERROR")'.format(xl_rowcol_to_cell(i, 2),
                                                                      xl_rowcol_to_cell(i, 3))

        worksheet.write(i, 0, item["TASK_ID"])
        worksheet.write(i, 1, item["TASK_STATUS"])
        worksheet.write(i, 2, str(actual_date))
        worksheet.write(i, 3, str(expected_date))
        worksheet.write(i, 4, formula)
        worksheet.write(i, 5, formula2)

    worksheet.conditional_format(5, 4, i, 5, {'type': 'text',
                                              'criteria': "containing",
                                              'value': "ERROR",
                                              'format': xlFormats['bad']})

    worksheet.add_table(4, 0, i, 5, { 'style': 'Table Style Light 13',
                                      'name': 'Table' + pm_id,
                                      'columns': [
                                          {'header': 'Task Count' if grouped else "Task ID"},
                                          {'header': 'Status'},
                                          {'header': 'Planned Start Date'},
                                          {'header': 'Expected Date'},
                                          {'header': 'Valid?'},
                                          {'header': 'Date Valid?'},
                                     ]})

    if verbosity >= 1:
        print("{} OK. {} Error. Total {}".format(ok_count, total - ok_count, total))

if __name__ == "__main__":
    main()

